design/ip-address-management-for-networkdata.md

```diff
@@ -0,0 +1,398 @@
+  errorMessage: ""
+```
+
+For each *IPClaim*, the controller reconciling the *IPCalim* will select an
+available IP address randomly from the *IPPool*, if the object name is not in
+the *allocations* map in the object *spec* or in the *addresses* map in the
+*status*, in that case it would select that IP address. An error would be
+reflected on the claim.
+
+Once the IP address is selected, the controller will create an *IPAddress*
+for that address. In case of conflict, it will list all the *IPAddress*
+objects that have an owner reference to this *IPPool* and update the
+status with the mapping of IP addresses and *IPPool* object names. It will
+then randomly select an available IP address. Once the *IPAddress* object
+is created, the *IPPool* object status will be updated with the new map. Then
+The *IPClaim* status will be updated with the *IPAddress* reference and the
+*ready* field set to true.
+
+If the *lastUpdated* field of the *IPPool* is empty, the controller will
+list all the *IPAddress* objects that have an owner reference to this
+*IPPool* and update the status with the mapping of IP addresses and
+*RenderedData* object names. It will also update the *lastUpdated* field.
+
+If the cluster to which the *IPClaim* belongs is paused, the reconciliation
+of the *IPClaim* would be paused.
+
+The *Template* would be modified this way:
+
+```yaml
+apiVersion: metadata.metal3.io/v1alpha1
+kind: Template
+metadata:
+  name: nodepool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3Cluster
+    name: cluster-1
+spec:
+  metaData:
+    ipAddressesFromIPPool:
+      - Name: pool-1
+        Key: "ip-address-1"
+    prefixesFromIPPool:
+      - Name: pool-1
+        Key: "netmask-1"
+    gatewaysFromIPPool:
+      - Name: pool-1
+        Key: "gateway-1"
+  networkData:
+    networks:
+      ipv4:
+        - id: "Baremetal"
+          link: "vlan1"
+          ipAddressfromIPPool: pool-1
+          routes:
+            - network: "0.0.0.0"
+              prefix: 0
+              gateway:
+                fromIPPool: pool-1
+              services:
+                - type: "dns"
+                  address: "8.8.4.4"
+        - id: "Provisioning"
+          link: "vlan2"
+          ipAddressfromIPPool: pool-2
+          routes:
+            - network: "0.0.0.0"
+              prefix: 0
+              gateway:
+                string: "192.168.1.1"
+              services:
+                - type: "dns"
+                  address: "8.8.4.4"
+status:
+  indexes:
+    "0": "machine-1"
+  dataNames:
+    "machine-1": nodepool-1-0
+  lastUpdated: "2020-04-02T06:36:09Z"
+```
+
+When reconciling the *RenderedData* object, the reconciler would create an
+*IPClaim* for the *IPPool* objects that are referenced in the
+*Template*. It will then wait until the *IPClaim* has a status set to
+Ready. Once ready, it will fetch the *IPAddress* and it will render the
```

Comment:
It could just wait for the ipAddress to be filled in. It could optionally look at errorMessage and relay that in its own status somehow.

Reply:
that's true, I will remove the flags